src/Components/WebAssembly/WebAssembly/src/Hosting/RootComponentTypeCache.cs

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Components.WebAssembly.Hosting
+{
+    internal static class RootComponentTypeCache
+    {
+        private static readonly Dictionary<Key, Type> _typeToKeyLookUp = new ();
+
+        public static Type GetRootComponent(string assembly, string type)
+        {
+            var key = new Key(assembly, type);
+            if (_typeToKeyLookUp.TryGetValue(key, out var resolvedType))
+            {
+                return resolvedType;
+            }
+            else
+            {
+                resolvedType = ResolveType(key, AppDomain.CurrentDomain.GetAssemblies());
+                _typeToKeyLookUp[key] = resolvedType;
+                return resolvedType;
+            }
+        }
+
+        private static Type ResolveType(Key key, Assembly[] assemblies)
+        {
+            var assembly = assemblies
+                .FirstOrDefault(a => string.Equals(a.GetName().Name, key.Assembly, StringComparison.Ordinal));
+
+            if (assembly == null)
+            {
+                return null;
+            }
+
+            return assembly.GetType(key.Type, throwOnError: false, ignoreCase: false);
+        }
+
+        private struct Key : IEquatable<Key>
```

Comment:
readonly struct? It feels weird for this to be a mutable struct.

Reply:
made this readonly in the shared implementation

Reply:
made this readonly on the unified cache type